### PR TITLE
Implement multi-core fuzzer

### PR DESCRIPTION
**Pull request by devatwork**
Refactor the fuzzer implementation into the engine and the
corpus_based_fuzzer components. This allowes multiple fuzzers
to co-exist side-by-side.

The engine is overhauled to support running on multiple core in
parallel. Big shoutout to @mvanbutselaar for his help with the
implementation. The output of the engine is colorized for improved
readability.

### CODE REVIEW

**Comment by DrRataplan** [comment]
## BundleMon

  
  
  
  
  
  
  
<details>
<summary>Unchanged files (2)</summary>

Status | Path | Size | Limits
:------------: | ------------ | :------------: | :------------:
:white_check_mark: | fontoxpath.esm.js<br/> | 88.68KB | -
:white_check_mark: | fontoxpath.js<br/> | 88.58KB | -
</details>



No change in files bundle size

Final result: :white_check_mark:

---
<p align="center"><a href="https://bundlemon.now.sh/projects/609024effb87a20009a6b096/reports?branch=threaded-fuzzer&resolution=all" target="_blank" rel="noreferrer noopener">Current branch size history</a> | <a href="https://bundlemon.now.sh/projects/609024effb87a20009a6b096/reports?branch=master&resolution=days" target="_blank" rel="noreferrer noopener">Target branch size history</a></p>

**Comment by devatwork** [comment]
@DrRataplan Not sure how to deal with the 2 remaining CodeFactor issues. @mvanbutselaar and I needed to alias the `Worker` class to `ThreadWorker`, so in my opinion both the casing is correct (because it is a type alias for a class) and the use of the spread operator for the aliasing. Any suggestions?

**Comment by DrRataplan** [comment]
I guess a big part of why the code is not deemed up-to-par is because of the mixed `import` and `require`. Using more `import as` should make this better. Did you already try that?

File: package-lock.json [hunk 1]
```diff
@@ -1,8372 +1,8 @@
 {
```
Comment (DrRataplan) [inline package-lock.json:1]: No need to include this file in the commit. You changed no dependencies.

**Comment by devatwork** [comment]
Thanks, my linter is seriously fubar and my TS skills are not up to par yet. Thanks for making the changes and maybe at some point we can jointly fix my linter because these issues waist both our time.

File: package-lock.json (hunk 1 above)
Comment (devatwork) [reply package-lock.json:1]: I remember installing `tsconfig-paths` but from the looks of it, it was already a dependency in the first place. I'll drop the file from the PR.

File: fuzzers/engine.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+export default class Engine<TFuzzer extends IFuzzer> {
+	run(fuzzer: TFuzzer, filename: string): void {
+		if (isMainThread) {
+			this.run_main(filename);
+		} else {
+			this.run_worker(fuzzer);
+		}
+	}
+
+	private run_main(filename: string): void {
```
Comment (DrRataplan) [inline fuzzers/engine.ts:24]: Let's keep to the codestyle and call this runMain. We do not really use underscores in this repo

**Review by DrRataplan — APPROVED** [review]

File: fuzzers/engine.ts (hunk 2 above)
Comment (devatwork) [reply fuzzers/engine.ts:24]: You're right, sloppy, too much Rust lately ;)

**Pull request merged**